src/NHibernate/Criterion/ContainsExpression.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Linq;
+using NHibernate.Dialect;
+using NHibernate.Engine;
+using NHibernate.SqlCommand;
+
+namespace NHibernate.Criterion
+{
+	[Serializable]
+	public class ContainsExpression : AbstractCriterion
+	{
+		private readonly object value;
+		private readonly IProjection projection;
+		private readonly TypedValue typedValue;
+		private readonly bool freetext;
+
+		public ContainsExpression(IProjection projection, object value, bool freetext)
+		{
+			this.projection = projection;
+			this.value = value;
+			typedValue = new TypedValue(NHibernateUtil.String, this.value, false);
+			this.freetext = freetext;
+		}
+
+		public ContainsExpression(string propertyName, object value, bool freetext)
+			: this(Projections.Property(propertyName), value, freetext)
+		{
+		}
+
+		public override SqlString ToSqlString(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			var columns = CriterionUtil.GetColumnNamesAsSqlStringParts(projection, criteriaQuery, criteria);
+			var value = criteriaQuery.NewQueryParameter(typedValue).Single();
+			var arguments = new[] { columns[0], value };
+			var functionName = GetFunctionName(criteriaQuery);
+
+			var dialectFunction = criteriaQuery.Factory.SQLFunctionRegistry.FindSQLFunction(functionName);
+			if (dialectFunction == null)
+			{
+				throw new HibernateException(string.Format("The current dialect '{0}' doesn't support the function: '{1}'",
+					criteriaQuery.Factory.Dialect, functionName));
+			}
+			return dialectFunction.Render(arguments, criteriaQuery.Factory);
+		}
+
+		private string GetFunctionName(ICriteriaQuery criteriaQuery)
+		{
+			var dialect = criteriaQuery.Factory.Dialect;
+
+			if (dialect is MsSql2000Dialect)
```

Comment:
Asking for type is a code smell. Not that it doesn't happen elsewhere in NHibernate, but I'd rather not introduce more of it.

Reply:
What do you suggest to do instead?

Reply:
Yes, it's certainly smelly there too :) But I'd suggest just try checking for whether a commonly named function is registered in the dialect or not.

Reply:
Thanks for the constructive discussion.

Checking for a function by name would mean searching for at least two possible flavors, "contains" & "match". For my purposes, I would also need to look for "freetext" for SQL Server. To me that sounds more fragile than type checking.

There's also a potential future clash in some RDBMs, e.g. [Oracle Text](https://docs.oracle.com/cd/B28359_01/text.111/b28303/query.htm#i1007233) has both "contains" & "matches" keywords.

Reply:
The name doesn't have to be "contains" or "match" (see below discussion about the ambiguity). Sure, it has an element of fragility, but type checking is completely non extensible. Consider if you'd like to add this support in a different dialect. Having to modify ContainsExpression to do that would stink. I would hope that at some point, "NHibernate.Core" doesn't even know what MsSql2000Dialect is. It just knows about dialects or whatever similar construct may take its place in the future. 

As I mentioned, this kind of smell is not uncommon, but I'd rather add big TODO:s to fixing them, rather than adding more.

Reply:
Of course it doesn't *have* to be "contains" or "match", but the name still has to be familiar to someone coming from the DB world. Also, I'm sure the major players put some serious though into it as well, and the more I think about it, the more I find "contains" to best fit the job, in that it hints about tokenization. However, an ORM connects two worlds and it's addressing / facing programmers, where indeed Contains from String becomes debatable.

With all that in mind, I think your final proposal below with "FullTextContains" fits best. Good idea. I'll rename after agreeing on the rest below.

For the name checking vs. my question above ("Add a virtual property in the Dialect base class for the function name?"), how about meeting in the middle, by defining an abstract Dialect.FullTextFunction string property, then have the actual dialects implement it? Or, to not have all the custom dialects out there complain during compilation after upgrade, the property could be non-abstract and equal to an empty string by default or null, overridden by the dialects that care about it. And then, the FullTextContainsExpression will check it for !IsEmptyOrNull.

Reply:
Done both & committed, please re-review.